include/rmm/mr/device/arena_memory_resource.hpp

```diff
@@ -235,7 +235,26 @@ class arena_memory_resource final : public device_memory_resource {
       }
     }
 
-    if (!global_arena_.deallocate(ptr, bytes)) { RMM_FAIL("allocation not found"); }
+    if (!global_arena_.deallocate(ptr, bytes)) {
+      // It's now possible to use per thread default streams along with another pool of streams.
+      // This means that it's possible for an arena to move from a thread or stream arena back
+      // into the global arena during a defragmentation and then move down into another arena
+      // type. For instance, thread arena -> global arena -> stream arena. If this happens and
+      // there was an allocation from it while it was a thread arena, we now have to check to
+      // see if the allocation is part of a stream arena.
+      // Only do this in exceptional cases to not affect performance and have to check all
```

Comment:
question: By "exceptional" I guess you mean, "deallocating from the arena I was expecting, or the global arena failed"?

Reply:
It means not the normal path. In this case the other choice was to change the above code to always check all the arenas, but that could potentially have a performance impact (I'm honestly not sure on that). So to minimize that only do the additional checks in the case we would have thrown the exception before.  In this particular case the arena moved from thread arena -> global -> stream arena, code previously wasn't checking the stream arena, now this code does.

Reply:
Thanks.